libnmstate/nm/active_connection.py

```diff
@@ -20,162 +20,369 @@
+                NM.DeviceState.IP_CONFIG
+                <= nm_dev.get_state()
+                <= NM.DeviceState.ACTIVATED
+            )
 
-class ActiveConnection:
-    def __init__(self, context=None, nm_ac_con=None):
-        self._ctx = context
-        self._act_con = nm_ac_con
+    return False
 
-        nmdevs = None
-        if nm_ac_con:
-            nmdevs = nm_ac_con.get_devices()
-        self._nmdev = nmdevs[0] if nmdevs else None
 
-    def import_by_device(self, nmdev=None):
-        assert self._act_con is None
+def is_activating(nm_ac, nm_dev):
+    if not nm_ac or not nm_dev:
+        return True
+    if nm_dev.get_state_reason() == NM.DeviceStateReason.NEW_ACTIVATION:
+        return True
 
-        if nmdev:
-            self._nmdev = nmdev
-        if self._nmdev:
-            self._act_con = self._nmdev.get_active_connection()
+    return (
+        nm_ac.get_state() == NM.ActiveConnectionState.ACTIVATING
+    ) and not is_activated(nm_ac, nm_dev)
 
-    def deactivate(self):
-        """
-        Deactivating the current active connection,
-        The profile itself is not removed.
 
-        For software devices, deactivation removes the devices from the kernel.
-        """
-        act_connection = self._nmdev.get_active_connection()
-        if (
-            not act_connection
-            or act_connection.props.state
-            == NM.ActiveConnectionState.DEACTIVATED
-        ):
+class ProfileActivation:
+    def __init__(self, ctx, iface_name, iface_type, nm_profile, nm_dev):
+        self._ctx = ctx
+        self._iface_name = iface_name
+        self._iface_type = iface_type
+        self._nm_ac = None
+        self._nm_dev = nm_dev
+        self._nm_profile = nm_profile
+        self._ac_handlers = set()
+        self._dev_handlers = set()
+        self._action = None
+
+    def run(self):
+        specific_object = None
+        self._action = (
+            f"Activate profile uuid:{self._nm_profile.get_uuid()} "
+            f"iface:{self._iface_name} type: {self._iface_type}"
+        )
+        if self._nm_dev:
+            # Workaround of https://bugzilla.redhat.com/1880420
+            mark_device_as_managed(self._ctx, self._nm_dev)
+
+        user_data = None
+        self._ctx.register_async(self._action)
+        self._ctx.client.activate_connection_async(
+            self._nm_profile,
+            self._nm_dev,
+            specific_object,
+            self._ctx.cancellable,
+            self._activate_profile_callback,
+            user_data,
+        )
+
+    @staticmethod
+    def wait(ctx, nm_ac, nm_dev):
+        obj = ProfileActivation(
+            ctx,
+            nm_dev.get_iface(),
+            get_iface_type(nm_dev),
+            None,
+            nm_dev,
+        )
+        obj._nm_ac = nm_ac
+        obj._action = (
+            f"Waiting activation of {obj._iface_name} {obj._iface_type}"
+        )
+        ctx.register_async(obj._action)
+        obj._wait_profile_activation()
```

Comment:
What about using `activation_handler` or a better name instead of `obj`? Thanks!

Reply:
Changed to `activation`, even through I like `obj` better. :)

Reply:
Thanks!